nearswap/src/internal.rs

```diff
@@ -28,22 +28,12 @@ impl NearSwap {
     /// in the pool is `in_bal` and `out_bal` of paid tokens and buying tokens respectively.
     #[inline]
     pub(crate) fn calc_out_amount(&self, in_amount: u128, in_bal: u128, out_bal: u128) -> u128 {
-        // in_a * out_bal / (in_bal + in_a)  and scaling for fee
-        let in_net = u256::from(in_amount) * 997;
-        let r: u256 = in_net * u256::from(out_bal) / (u256::from(in_bal) * 1000 + in_net);
-        return r.as_u128();
-    }
+        // formula r = (in_amount * in_bal * out_bal ) / ( in_amount + in_bal ) ^ 2;
+        let denominator = (U512::from(in_amount) + U512::from(in_bal)) * U512::from(in_amount) + U512::from(in_bal);
+        let numerator = ( U512::from(in_amount) * U512::from(in_bal) * U512::from(out_bal));
```

Comment:
We don't need to use `U512`. Multiplying 2 `U128` numbers will never make a number bigger than `U256`.

`x * Y * X` could make a 512 bit number - but this won't happen, as the pool would need to consume more than a half of the liquidity of Y and X, and x will need to be an order of X.

Reply:
ok